### DISCUSSION TITLE
access newly created record from within a nested write (update first, then create within)

### DISCUSSION THREAD

**Discussion by miller-productions** [prisma/prisma#4788] [Q&A] [answered]
Hi there,

I have a question about accessing a record that was newly created from within a nested write (update first, then create within).

I'm following along the examples on [this page in the prisma docs](https://www.prisma.io/docs/concepts/components/prisma-client/relation-queries).

In particular, I am looking at the following two items in the data model:

Note, I have slightly modified it for the purposes of this question by adding a `counter` to `User`.

```prisma
model User {
  id       Int     @id @default(autoincrement())
  email    String  @unique
  posts    Post[]
  counter  Int
}
model Post {
  id         Int      @id @default(autoincrement())
  title      String
  author     User?    @relation(fields:  [authorId], references: [id])
  authorId   Int?
}
```

Now, let's say you want to create a new `Post` and connect it to the `User` and at the same time, you also want to increment the `counter`.

My assumption, after reading [this section of the doc](https://www.prisma.io/docs/concepts/components/prisma-client/relation-queries#update-an-existing-user-record-by-creating-a-new-post-record) is that you would need to update the existing `User` and within that, create a new `Post` record.

i.e.

```javascript
const user = await prisma.user.update({
  where: { email: 'alice@prisma.io' },
  data: {
    // increment the counter for this User
    counter: {
      increment: 1,
    },
    // create the new Post for this User
    posts: {
      create: { title: 'Hello World' },
    },
  },
})
```

My question is this. In the above scenario, how do you access the newly created `Post` in the return of the query?

In particular, say you want to get the `id` of the new `Post`?

**Comment by miller-productions** [comment]
I have found one way of achieving this that works.

It uses [the `tranaction` api](https://www.prisma.io/docs/concepts/components/prisma-client/transactions#the-transaction-api).

Basically, instead of trying to use a nested write that begins with an update on `User` and contains a nested create on `Post`...

I split the two operations out into separate variables and run them together using a `prisma.$transaction()` as follows.

```javascript
const updateUser = prisma.user.update({
  where: { email: 'alice@prisma.io' },
  data: {
    // increment the counter for this User
    counter: {
      increment: 1,
    },
  },
});

const createPost = prisma.post.create({
  data: {
    // create the new Post for this User
    title: 'Hello World',
    author: {
      connect: {
        email: 'alice@prisma.io',
      },
    },
  },
});

const [ updateUserResult, createPostResult ] = await prisma.$transaction([updateUser, createPost ]);

// if createPostResult, then can access details of the new post from here...
```

**Comment by nikolasburk** [comment, marked as answer]
Hey @miller-productions, great to see that you already found a solution for your problem! I just want to add that you can also use the [`include`](https://www.prisma.io/docs/concepts/components/prisma-client/field-selection#include) option in the nested write so retrieve the info about the `posts` relation. It would look similar to his:

```ts
const user = await prisma.user.update({
  where: { email: 'alice@prisma.io' },
  data: {
    // increment the counter for this User
    counter: {
      increment: 1,
    },
    // create the new Post for this User
    posts: {
      create: { title: 'Hello World' },
    },
  },
  include: { posts: true },
})
```

You can also combine `include` and `select` to only retrieve the `id` field of the posts:

```ts
const user = await prisma.user.update({
  where: { email: 'alice@prisma.io' },
  data: {
    // increment the counter for this User
    counter: {
      increment: 1,
    },
    // create the new Post for this User
    posts: {
      create: { title: 'Hello World' },
    },
  },
  include: {
    posts: {
      select: {
        id: true,
      },
    },
  },
})
```

Please let me know if that helps or if you have any further questions, always happy to help 🙌

**Reply by miller-productions** [reply]
Thanks @nikolasburk, your approach is what I was looking for. That solves it more elegantly.

**Comment by ethndotsh** [comment]
I too am having this issue. When you use `include`, it returns ALL of the existing relations, and all I want is the ID of the singular post I created. Any other suggestions on how I could accomplish this?

**Reply by jblyberg** [reply]
I do something like this:

```ts
const user = await prisma.user.update({
  where: { email: 'alice@prisma.io' },
  data: {
    // increment the counter for this User
    counter: {
      increment: 1,
    },
    // create the new Post for this User
    posts: {
      create: { title: 'Hello World' },
    },
  },
  include: {
    posts: {
      orderBy: {
        createdAt: 'desc',
      },
      take: 1,
    },
  },
})
```

**Reply by miller-productions** [reply]
Good point @cursecodes. 
Looks like @jblyberg's solution there is a candidate. It assumes we have `createdAt` as a field in our `posts` table to `orderBy`, and if we have that, this should enable us to get the post that we just created in the nested write.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
